### PR TITLE
feat: enable plugin-specific comment functionality

### PR DESCRIPTION
**Pull request by mgrip**
This change enables language plugins to provide their own "handleComment" function. To do so, we're exposing the addLeadingComment, addTrailingComment, and addDanglingComment functions which attach comment nodes correctly in the ast,  which will later get handled by prettier core when printing.

WIP - feedback and contributions are welcome

### CODE REVIEW

**Comment by mgrip** [comment]
Addresses https://github.com/prettier/prettier/issues/4177

File: src/common/util.js [hunk 1]
```diff
@@ -756,6 +756,37 @@ function arrayify(object, keyName) {
...
+  const comments = node.comments || (node.comments = []);
+  comments.push(comment);
+  comment.printed = false;
+
+  // For some reason, TypeScript parses `// x` inside of JSXText as a comment
+  // We already "print" it via the raw text, we don't need to re-print it as a
+  // comment
+  if (node.type === "JSXText") {
+    comment.printed = true;
+  }
```
Comment (vjeux) [inline src/common/util.js]: This should probably be inside of estree printer plugin

**Comment by mgrip** [comment]
@vjeux do you think we should move those estree-specific handlers out into a `language-js` `handleComment` function? and if so part of this PR or a follow-up?

**Comment by vjeux** [comment]
`attach` has some boilerplate that is common and then the bulk of the parser-specific logic is in all the `handleXXXX`.

https://github.com/prettier/prettier/blob/master/src/main/comments.js#L167-L193

If you look, there are three cases:

1)
```js
    if (privateUtil.hasNewline(text, locStart(comment), { backwards: true })) {
```

2) 
```js
    } else if (privateUtil.hasNewline(text, locEnd(comment))) {
```

3)
```js
    } else {
```

Ideally, plugins should be able to give in a list of handlers for those three passes but not have to reimplement the rest of the logic.

**Comment by vjeux** [comment]
As for this PR or not, I don't really care. This current PR seems to be working and provide value so we can merge it as is.

**Comment by mgrip** [comment]
Ahh ok that makes sense. So like `handleComment`, `handleCommentOnNewLine`, `handleCommentWithTrailingNewLine`? (Naming is hard 🙄 )

**Comment by vjeux** [comment]
The idea between the three cases are:

- Comment on its own line
- Comment at the end of a line
- The rest

**Comment by mgrip** [comment]
Ok got it. I'll add `handleComment`, `handleEndOfLineComment`, `handleCommentOnOwnLine`? And I can separate out the JS specific ones as part of this

**Comment by vjeux** [comment]
I think that it would be helpful is the interface was an array of handlers and have the generic code go through all the handlers one by one and stop if one returns false. The same way the logic is setup with lots of else ifs.

**Comment by mgrip** [comment]
Sorry not sure I understand completely - do you mean only call the plugin-specific handlers in case 3?

**Comment by vjeux** [comment]
The code looks like this:

```js
      if (
        handleLastFunctionArgComments(
          text,
          precedingNode,
          enclosingNode,
          followingNode,
          comment,
          options
        ) ||
        handleConditionalExpressionComments(
          enclosingNode,
          precedingNode,
          followingNode,
          comment,
          text,
          options
        ) ||
        handleImportSpecifierComments(enclosingNode, comment) ||
        handleIfStatementComments(
          text,
          precedingNode,
          enclosingNode,
          followingNode,
          comment,
          options
        ) ||
        handleClassComments(
          enclosingNode,
          precedingNode,
          followingNode,
          comment
        ) ||
        handleLabeledStatementComments(enclosingNode, comment) ||
        handleCallExpressionComments(precedingNode, enclosingNode, comment) ||
        handlePropertyComments(enclosingNode, comment) ||
        handleExportNamedDeclarationComments(enclosingNode, comment) ||
        handleOnlyComments(enclosingNode, ast, comment, isLastComment) ||
        handleTypeAliasComments(enclosingNode, followingNode, comment) ||
        handleVariableDeclaratorComments(enclosingNode, followingNode, comment)
```

Ideally the plugin would export something that looks like:

```js
handleComments: {
  ownLine: [
    handleLastFunctionArgComments,
    handleConditionalExpressionComments,
    handleImportSpecifierComments,
    ...
  ],
  endOfLine: [
    ...
  ],
  remaining: [
    ...
  ],
}
```

**Comment by mgrip** [comment]
Got it. The only thing I'm a little hesitant about is the array form sort of implies precedence, based on the order of things in the array. If `ownLine` is just a function then the plugin developer has to explicitly define precedence

**Comment by vjeux** [comment]
Thanks a lot for improving the internals of prettier :)

**Comment by vjeux** [comment]
I was going to suggest doing the same for locStart/locEnd but looks like it's already been done :) https://github.com/prettier/prettier/blob/13103101ebd6bc563fb6cc765186244b90d8300e/src/language-js/index.js#L11-L37

This was the last thing that I was aware needed to be plugin-ified

**Comment by mgrip** [comment]
I think this was the last thing we noticed too :) At this point I don't think there's much stopping us from getting to our first stable version for the php plugin! https://github.com/prettier/plugin-php/issues/181

**Comment by azz** [comment]
This is awesome, great work @mgrip!

**Pull request merged**